connector/connect/client/jvm/src/main/scala/org/apache/spark/ml/classification/Classifier.scala

```diff
@@ -0,0 +1,187 @@
+
+  override protected def validateAndTransformSchema(
+      schema: StructType,
+      fitting: Boolean,
+      featuresDataType: DataType): StructType = {
+    val parentSchema = super.validateAndTransformSchema(schema, fitting, featuresDataType)
+    SchemaUtils.appendColumn(parentSchema, $(rawPredictionCol), new VectorUDT)
+  }
+}
+
+/**
+ * Single-label binary or multiclass classification. Classes are indexed {0, 1, ..., numClasses -
+ * 1}.
+ *
+ * @tparam FeaturesType
+ *   Type of input features. E.g., `Vector`
+ * @tparam E
+ *   Concrete Estimator type
+ * @tparam M
+ *   Concrete Model type
+ */
+abstract class Classifier[
+    FeaturesType,
+    E <: Classifier[FeaturesType, E, M],
+    M <: ClassificationModel[FeaturesType, M]]
+    extends Predictor[FeaturesType, E, M]
+    with ClassifierParams {
+
+  /** @group setParam */
+  def setRawPredictionCol(value: String): E = set(rawPredictionCol, value).asInstanceOf[E]
+
+  // TODO: defaultEvaluator (follow-up PR)
+}
+
+/**
+ * Model produced by a [[Classifier]]. Classes are indexed {0, 1, ..., numClasses - 1}.
+ *
+ * @tparam FeaturesType
+ *   Type of input features. E.g., `Vector`
+ * @tparam M
+ *   Concrete Model type
+ */
+abstract class ClassificationModel[FeaturesType, M <: ClassificationModel[FeaturesType, M]]
+    extends PredictionModel[FeaturesType, M]
+    with ClassifierParams {
+
+  /** @group setParam */
+  def setRawPredictionCol(value: String): M = set(rawPredictionCol, value).asInstanceOf[M]
+
+  /** Number of classes (values which the label can take). */
+  def numClasses: Int
+
+  override def transformSchema(schema: StructType): StructType = {
+    var outputSchema = super.transformSchema(schema)
+    if ($(predictionCol).nonEmpty) {
+      outputSchema = SchemaUtils.updateNumValues(schema, $(predictionCol), numClasses)
+    }
+    if ($(rawPredictionCol).nonEmpty) {
+      outputSchema =
+        SchemaUtils.updateAttributeGroupSize(outputSchema, $(rawPredictionCol), numClasses)
+    }
+    outputSchema
+  }
+
+  /**
+   * Transforms dataset by reading from [[featuresCol]], and appending new columns as specified by
+   * parameters:
+   *   - predicted labels as [[predictionCol]] of type `Double`
+   *   - raw predictions (confidences) as [[rawPredictionCol]] of type `Vector`.
+   *
+   * @param dataset
+   *   input dataset
+   * @return
+   *   transformed dataset
+   */
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val outputSchema = transformSchema(dataset.schema, logging = true)
+
+    // Output selected columns only.
+    // This is a bit complicated since it tries to avoid repeated computation.
+    var outputData = dataset
+    var numColsOutput = 0
+    if (getRawPredictionCol != "") {
+      val predictRawUDF = udf { features: Any =>
+        predictRaw(features.asInstanceOf[FeaturesType])
+      }
+      // TODO: Enable when DataFrame.withColumn support metadata
```

Comment:
Q: When will spark connect supporting column metadata ?
Q2: How many cases in spark ML requires column metadata ?

Reply:
Q1: I just note that it is a [private method](https://github.com/apache/spark/blob/a1649ad24298d988267acb8588d19848c7fb16c4/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L2798), but should not be hard to add it.
Q2: IIRC, usually column metadata is only optional, it can be the `numClasses` or `numFeatures` so the underlying implementation may skip an action to obtain them.

Reply:
> numClasses

I remember if without `metadata`, we cannot support numClasses > 100, right ?

Reply:
yes.
do we still need this limitation maxNumClasses == 100?